odl/contrib/pytorch/operator.py

```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
 
             grad = torch.from_numpy(np.array(grad_odl, copy=False, ndmin=1))
 
+            if grad_output.is_cuda:
+                # Push back to GPU
+                grad = grad.cuda()
+
         return grad
 
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        return '{}(\n    {!r}    \n)'.format(self.__class__.__name__,
+                                             self.operator)
+
+
+class OperatorAsModule(torch.nn.Module):
+
+    """Wrapper of an ODL operator as a ``torch.nn.Module``.
+
+    This wrapper can be used as a layer in ``pytorch`` Neural Networks.
+    It works with arbitrary batches and channels and supports
+    backpropagation.
+
+    .. note::
+        Currently, batches and channels are supported by simply looping
+        over them and stacking the results.
+    """
+
+    def __init__(self, operator):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        operator : `Operator`
+            The ODL operator to be wrapped. For gradient computations to
+            work, ``operator.derivative(x).adjoint`` must be implemented.
+
+        Examples
+        --------
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype='float32')
+        >>> odl_op = odl.MatrixOperator(matrix)
+        >>> torch_op = OperatorAsModule(odl_op)
+        >>> torch_op.op_func.operator is odl_op
+        True
+        """
+        super(OperatorAsModule, self).__init__()
+        self.op_func = OperatorAsAutogradFunction(operator)
+
+    def forward(self, x):
+        """Compute forward-pass of this module on ``x``.
+
+        Parameters
+        ----------
+        x : `torch.autograd.variable.Variable`
+            Input of this layer. The contained tensor must have shape
+            ``extra_shape + operator.domain.shape``, and
+            ``len(extra_shape)`` must be at least 1 (batch axis).
```

Comment:
That would just be an extra axis. This module is a bit like a linear layer that only understands input shape, output shape and batch axis. But in contrast to the linear layer, which can literally only act with one extra axis, this one supports any number of extra axes (batch, channel, you name it), simply by flattening them and then looping. Since there are no parameters to learn, any extra axis looks the same.

Reply:
This made 100% sense, try to get some of that into the doc!